### ISSUE TITLE
alternator_tests.py::TesterAlternator::test_cluster_traces: KeyError: 'The following "accept_proposal: send accept proposal" trace message not found from "127.0.29.1" source!'

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#10189]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/1151/testReport/alternator_tests/TesterAlternator/test_cluster_traces/
Scylla version a4756334cee208f5ccb080df222af78dde6e1eb6

```
Error Message
KeyError: 'The following "accept_proposal: send accept proposal" trace message not found from "127.0.29.1" source!'
Stacktrace
self = <alternator_tests.TesterAlternator object at 0x7fa0cd9eb550>

    def test_cluster_traces(self):  # pylint:disable=too-many-locals,too-many-statements
        """
        The test inserts items of different types and checks the traces for each of the following actions: "PutItem",
         "GetItem", "UpdateItem", and "DeleteItem".
        Also, for each action taken, the test checks for the following:
         1. For each action, the test checks a list of trace messages that should exist.
         2. The order of the traces should be according to the order of the expected traces variable (In other words,
          the test checks for the first message, and once is found, the test continue to find for the next message.
          Until the test found all the messages in that order).
         3. Some of the traces are supposed to appear under each node.
        """
        table_name = TABLE_NAME
        num_of_items = 10
        self.prepare_dynamodb_cluster(num_of_nodes=3)
        nodes = self.cluster.nodelist()
        self.create_table(node=nodes[0], table_name=table_name)
        data_generator = AlternatorDataGenerator(
            primary_key=self._table_primary_key, primary_key_format=self._table_primary_key_format)
        items = data_generator.create_multiple_items(num_of_items=num_of_items, mode=TypeMode.MIXED)
        # This dictionary contains the "traces" messages for each method. Also, the order of messages is important!
        # During the test, we expect to find the messages in the order they are displayed in the list.
        # For example: For the "put_item" method, the test expects to see first the "PutItem" message.
        # The next message should be the "accept_proposal: send accept proposal" message.
        # Finally, the test expects to see "CAS successful" message (if one of those messages does not appear in
        #  this order, the test will fail).
        expected_messages_dict = {
            # Last thing, the numbers that appear next to a message indicate the number of different messages we are
            #  looking for. That is, if the number is not 1, we expect to see the message from several different nodes.
            'put': [('PutItem', 1), ('prepare_ballot: sending prepare', len(nodes) - 1),
                    ('accept_proposal: send accept proposal', len(nodes) - 1), ('CAS successful', 1)],
            'get': [('GetItem', 1), ('Creating read executor for token', 1), ('Querying is done', 1)],
            'update': [('UpdateItem', 1), ('accept_proposal: send accept proposal', len(nodes) - 1),
                       ('prune: send prune of', len(nodes) - 1), ('CAS successful', 1)],
            'delete': [('DeleteItem', 1), ('accept_proposal: send accept proposal', len(nodes) - 1),
                       ('prune: send prune of', len(nodes) - 1), ('CAS successful', 1)],
        }
        method_name_by_method_idx = {method_idx: method_name
                                     for method_idx, method_name in enumerate(expected_messages_dict)}
        expected_traces_number = len(expected_messages_dict)
    
        set_trace_probability(nodes=nodes, probability_value=1.0)
        for item_idx, item in enumerate(items):
            node = nodes[item_idx % len(nodes)]
            item_key = {self._table_primary_key: item[self._table_primary_key]}
            for method_name in expected_messages_dict:
                if method_name == 'put':
                    self.put_item(node=node, item=item, table_name=table_name)
                elif method_name == 'get':
                    self.get_item(node=node, item_key=item_key, table_name=table_name, consistent_read=True)
                elif method_name == 'update':
                    self.update_item(node=node, item_key=item_key, table_name=table_name)
                elif method_name == 'delete':
                    self.delete_item(node=node, item_key=item_key, table_name=table_name)
                else:
                    raise KeyError(f'The following "{method_name}" method name not supported!')
    
        set_trace_probability(nodes=nodes, probability_value=0.0)
        expected_traces_size = len(items) * expected_traces_number
        logger.info(f'Expecting to find at least "{expected_traces_size}" partitions')
        # For each method we use we get "len(items)" traces. Therefore, in our case we used 2
        # (len(expected_messages_dict)) methods and 10 (len(item)) items.
        # Therefore we will observe "len(items) * methods_size" messages.
        all_traces_events = self.get_all_traces_events(expected_traces_size=expected_traces_size)
        node_ips = {get_ip_from_node(node) for node in nodes}
        # The following action order for each item is: "PutItem", "GetItem", "UpdateItem", and "DeleteIte" (this
        #  order is order of "expected_messages_dict" keys).
        # Therefore, for each action, we need to get a list with "len(items)" cells.
        events_by_action = {}
        for events_idx, events in enumerate(all_traces_events):
            events_by_action.setdefault(
                method_name_by_method_idx[events_idx % expected_traces_number], []).append(events)
    
        def verify_traces_messages(method_name):  # pylint:disable=too-many-locals
            all_traces = events_by_action[method_name]
            expected_messages = expected_messages_dict[method_name]
            # The "all_traces" variable contains the list of traces in the order of action ("get_item" or "pu_item")
            #  we did. The test enters multiple items. Thus, need over on the traces for each item entered.
            for action_idx, traces in enumerate(all_traces):
                source = get_ip_from_node(nodes[action_idx % len(nodes)])
                trace_idx = 0
                # This loop goes over the messages that should appear within the traces in the order of the
                # "expected_messages".
                for expected_message_details in expected_messages:
                    result = []
                    expected_message, expected_message_number = expected_message_details
                    count = expected_message_number
                    is_message_found = False
                    # This loop goes through all the traces once and tries to find the expected messages.
                    # Therefore, if the message not found or the while loop ends, this is means that a test failed
                    #  because the messages order was incorrect or the expected message changed.
                    while trace_idx < len(traces):
                        trace = traces[trace_idx]
                        trace_idx += 1
                        event_msg = trace['activity']
                        if expected_message in event_msg and source == trace['source']:
                            if expected_message != event_msg:
                                result.append(event_msg)
                            count -= 1
                            if count == 0:
                                is_message_found = True
                                break
                    if not is_message_found:
                        raise KeyError(f'The following "{expected_message}" trace message not found from "{source}"'
                                       f' source!')
                    if result:
                        if method_name in ['put', 'update', 'delete']:
                            ips = {msg.rsplit(' ', maxsplit=1)[1] for msg in result}
                            _diff = (node_ips ^ ips)
                            assert _diff == {source}, \
                                f'The "{expected_message}" message not sent from "{expected_message_number}" ' \
                                f'nodes (The message in missing in the following "{_diff}" IPs'
                        elif method_name == 'get':
                            for msg in result:
                                ips = set(node_ip.strip() for node_ip in
                                          msg.split('{', maxsplit=1)[1].split('}', maxsplit=1)[0].split(','))
                                _diff = (node_ips ^ ips)
                                assert not _diff, f'The following node ips "{_diff}" not found in "get_item" event'
                        else:
                            raise KeyError(f'The following "{method_name}" method name not supported!')
    
        for method_name in expected_messages_dict:
            logger.info(f'Verifying all traces of "{method_name}_item" method name')
>           verify_traces_messages(method_name=method_name)

alternator_tests.py:952: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

method_name = 'delete'

    def verify_traces_messages(method_name):  # pylint:disable=too-many-locals
        all_traces = events_by_action[method_name]
        expected_messages = expected_messages_dict[method_name]
        # The "all_traces" variable contains the list of traces in the order of action ("get_item" or "pu_item")
        #  we did. The test enters multiple items. Thus, need over on the traces for each item entered.
        for action_idx, traces in enumerate(all_traces):
            source = get_ip_from_node(nodes[action_idx % len(nodes)])
            trace_idx = 0
            # This loop goes over the messages that should appear within the traces in the order of the
            # "expected_messages".
            for expected_message_details in expected_messages:
                result = []
                expected_message, expected_message_number = expected_message_details
                count = expected_message_number
                is_message_found = False
                # This loop goes through all the traces once and tries to find the expected messages.
                # Therefore, if the message not found or the while loop ends, this is means that a test failed
                #  because the messages order was incorrect or the expected message changed.
                while trace_idx < len(traces):
                    trace = traces[trace_idx]
                    trace_idx += 1
                    event_msg = trace['activity']
                    if expected_message in event_msg and source == trace['source']:
                        if expected_message != event_msg:
                            result.append(event_msg)
                        count -= 1
                        if count == 0:
                            is_message_found = True
                            break
                if not is_message_found:
>                   raise KeyError(f'The following "{expected_message}" trace message not found from "{source}"'
                                   f' source!')
E                   KeyError: 'The following "accept_proposal: send accept proposal" trace message not found from "127.0.29.1" source!'

alternator_tests.py:932: KeyError
```

I don't know if the issue is in scylla or in the dtest (or both :)).
@nyh can you please look into this?

**Comment by nyh** [comment]
Do you know if it fails every time, or it's just a fluke?
If it's every time, maybe that magical trace message "accept_proposal: send accept proposal" is no longer printed? 

In any case, I think this test is too detailed. The idea was to check that tracing was vaguely working - not that the details of how LWT writes trace messages never change.

Seeing that the author of this test is no longer in Scylla, I can try to run this test myself and see what's going on.

**Comment by nyh** [comment]
The test passed on my laptop - I tried running it 5 times (each time takes ages because my dtest speedup patch isn't in) and it passed each time... 

According to https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/1151/testReport/alternator_tests/TesterAlternator/test_cluster_traces/history/

This test failed 3 times in the last 12 builds. So it's not a one time thing, but also not a reliable failure but perhaps some kind of race condition.

Unfortunately, the test is written in a way that it counts the different trace messages, and prints a generic error message if one of the counts is wrong - I have no way to tell *why* it failed, or what did the traces looked like when it failed.
So perhaps the only thing we can reasonably do now is to write this test to print out the entire traces that it reads, to help debugging (I'm not sure, however, where we'll see the printouts - I don't see Jenkins showing any log messages beyond the error messages...)

**Comment by nyh** [comment]
I'm sorry, I looked at the test, and I simply don't understand what it does :-( It has layers upon layers of variables and abstractions, and I'm completely lost in the details to figure out what went wrong - certainly when I can't reproduce the failure myself... I don't know if the trace message "accept_proposal: send accept proposal" doesn't appear on one of the nodes, or appears in the wrong order, or perhaps does appear and just the counting code is wrong.

I see in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/1151/testReport/alternator_tests/TesterAlternator/test_cluster_traces/history/ that unfortunately it is not a one-time fluke - it failed 5 times out of the last 25 runs and the last failure was two days ago :-( Maybe I'll just to run it more than 5 times to reproduce it... If I could reproduce it I could add printouts to understand what's going on.

**Comment by nyh** [comment]
This test has continued to fail intermittently - see duplicate issue #11874 for newer evidence.

**Comment by nyh** [comment]
@kostja @gleb-cloudius I have an LWT question for you. This dtest does some Alternator operations which use LWT, and then looks for specific things in the traces (which I think is not a good idea, but this what this dtest does). One of the strings it looks for in each trace is **`accept_proposal: send accept proposal`**, and rarely (I couldn't reproduce this myself, but it does fail on Jenkins sometimes) this string is missing in the log.

So I have three simple (I hope) LWT questions for you:
1. I'm not sure what "accept proposal" means. The test seems to demand that if we have N nodes, we will see "accept proposal" on N-1 on them. The is there a possibility that once in a while, because of timing or something, LWT doesn't do an "accept proposal" and does something else instead?
2. I see the log message is "accept_proposal: send accept {}" with "{}" being the variable proposal, not the string "proposal". Does this always print as the string "proposal" as the test expects? Are other other possibilities?
3. One of the tests expects to see first a `prepare_ballot: sending prepare` message, and only then the `accept_proposal: send accept proposal`. Is it possible that in some rare cases this order can be reversed?

Thanks.

**Comment by gleb-cloudius** [comment]
> @kostja @gleb-cloudius I have an LWT question for you. This dtest does some Alternator operations which use LWT, and then looks for specific things in the traces (which I think is not a good idea, but this what this dtest does). One of the strings it looks for in each trace is **`accept_proposal: send accept proposal`**, and rarely (I couldn't reproduce this myself, but it does fail on Jenkins sometimes) this string is missing in the log.
> 
> So I have three simple (I hope) LWT questions for you:
> 
>     1. I'm not sure what "accept proposal" means. The test seems to demand that if we have N nodes, we will see "accept proposal" on N-1 on them. The is there a possibility that once in a while, because of timing or something, LWT doesn't do an "accept proposal" and does something else instead?

The proposal is sent by a coordinator to each live paxos participant but itself. So the test expects all endpoints to be alive and participate in a paxos round.

> 
>     2. I see the log message is "accept_proposal: send accept {}" with "{}" being the variable proposal, not the string "proposal". Does this always print as the string "proposal" as the test expects? Are other other possibilities?

Yes. This is how its 'operator<<' is implemented.

> 
>     3. One of the tests expects to see first a `prepare_ballot: sending prepare` message, and only then the `accept_proposal: send accept proposal`. Is it possible that in some rare cases this order can be reversed?
> 
No. This is paxos stages. First you do prepare than accept. Of course you need to be sure it is done for the same ballot.

Are you sure the logger does not have rate limiter and just eats messages sometimes?

**Comment by nyh** [comment]
@gleb-cloudius thanks. It's not "the logger" in this case, it's tracing of the request, and I assume nothing gets eaten there.
I don't know how to proceed from here. I think I'll start by trying harder to reproduce this failure locally...

**Comment by gleb-cloudius** [comment]
On Mon, Nov 07, 2022 at 12:19:35AM -0800, nyh wrote:
> @gleb-cloudius thanks. It's not "the logger" in this case, it's tracing of the request, and I assume nothing gets eaten there.
> I don't know how to proceed from here. I think I'll start by trying harder to reproduce this failure locally...
> 

Tracing is "eventually consistent" as well. Writes to the trace table can be dropped.


--
			Gleb.

**Comment by nyh** [comment]
Thanks @gleb-cloudius I think you hit the nail right on the head. It was obvious for me that this was handled, but your comment forced me to look at exactly how, and it seems it was handled wrong:

Obviously, in the test that I wrote for tracing, I made sure to retry for a while because it might take time (and in fact, usually takes a substantial amount of time) for all the traces to reach the disk.
But in this dtest, the code was written (not by me) in so many layers of utility functions, that it was hard to tell what retrying was going on. It turns out that the get_all_traces_events() function claims it retries - but it doesn't wait enough! It takes a number of expected events, and only waits until there is this number of events. But this number ("get_all_traces_events") had nothing to do with the real number of expected events - it was just the count of different operations and nodes - but we may expect more than one event from each node, and don't wait for it.

Anyway, I now believe that **this is a test bug.**

In my opinion, the test shouldn't try to guess the number of "expected" entries. Rather it needs to do what I did in test/alternator/test_tracing.py: repeatedly check if the trace contains what we hope it to contain - until a timeout. On error - we need to wait until the timeout - we can't tell if we saw "enough" other trace messages and the one we expected is really missing.

I'm un-assigning myself from this issue. Let's get someone from QA to fix the test (I'll be happy to explain how/why).
